h2/src/main/org/h2/engine/Session.java

```diff
@@ -162,6 +163,12 @@
     private State state = State.INIT;
     private long startStatement = -1;
 
+    /**
+     * Set of database object ids to be released at the end of transaction
+     */
+    private final BitSet idsToRelease = new BitSet();
```

Comment:
May be create this set only when required? Most similar fields in `Session` are created only on their first usage. Number of `Session` object can be very large, but this field is rarely needed.

Reply:
The only reason I did it this way, was for the "final", but you are right - I should follow the trend of lazy allocation here. 
Realistically speaking ```"Number of Session object can be very large" is an overstatement. ```, H2 will choke long before it reaches thousands of sessions. Of course one can create gazillion of idle ones, but if it would be even some light activity on 10% of them...

More importantly, I just realized that ids should not be released in case of a rollback, except for temp tables.